util/conversion/conversion.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conversion
+
+import (
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/json"
+)
+
+const (
+	DataAnnotation = "cluster.x-k8s.io/conversion-data"
+)
+
+var (
+	ErrDataNotFound = errors.New("not found")
+)
+
+// MarshalData stores the source object as json data in the destination object annotations map.
+func MarshalData(src metav1.Object, dst metav1.Object) error {
+	data, err := json.Marshal(src)
+	if err != nil {
+		return err
+	}
+	if dst.GetAnnotations() == nil {
+		dst.SetAnnotations(map[string]string{})
+	}
+	dst.GetAnnotations()[DataAnnotation] = string(data)
+	return nil
+}
+
+// UnmarshalData tries to retrieve the data from the annotation and unmarshals it into the object passed as input.
+func UnmarshalData(from metav1.Object, to interface{}) error {
+	data, ok := from.GetAnnotations()[DataAnnotation]
+	if !ok {
+		return ErrDataNotFound
+	}
+	if err := json.Unmarshal([]byte(data), to); err != nil {
```

Comment:
Would this properly handle an annotation that was set with a previous version? i.e. When we upgrade to v1alpha4, and there was an existing annotation on the resource with the v1alpha3 type, would this properly decode the v1alpha4 type from the v1alpha3 type embedded in the annotation?

Reply:
hmm, that's a good point. I think we should avoid persisting the annotation when Unmarshalling, so when we go from v1a2 -> v1a3 (or whatever the storage version is), we delete the annotation and data from the map, how does that sound?

We can do this operation as part of the `UnmarshalData` function

Reply:
sgtm

Reply:
done